# Lab1-Assignment

Copyright: Vrije Universiteit Amsterdam, Faculty of Humanities, CLTL

This notebook describes the assignment for Lab 1 of the text mining course. 

**Points**: each exercise is prefixed with the number of points you can obtain for the exercise.

We assume you have worked through the following notebooks:
* **Lab1.1-introduction**
* **Lab1.2-introduction-to-NLTK**
* **Lab1.3-introduction-to-spaCy** 

In this assignment, you will process an English text (**Lab1-apple-samsung-example.txt**) with both NLTK and spaCy and discuss the similarities and differences.

## Credits
The notebooks in this block have been originally created by [Marten Postma](https://martenpostma.github.io). Adaptations were made by [Filip Ilievski](http://ilievski.nl).

## Tip: how to read a file from disk
Let's open the file **Lab1-apple-samsung-example.txt** from disk.

In [29]:
from pathlib import Path

In [30]:
cur_dir = Path().resolve() # this should provide you with the folder in which this notebook is placed
path_to_file = Path.joinpath(cur_dir, 'Lab1-apple-samsung-example.txt')
print(path_to_file)
print('does path exist? ->', Path.exists(path_to_file))

C:\Users\nicol\Documents\Group-4-Text-Mining\lab_sessions\lab1\Lab1-apple-samsung-example.txt
does path exist? -> True


If the output from the code cell above states that **does path exist? -> False**, please check that the file **Lab1-apple-samsung-example.txt** is in the same directory as this notebook.

In [62]:
# added encoding utf 8 to open call because it was getting misread by windows
with open(path_to_file, encoding='utf-8') as infile:
    text = infile.read()

print('number of characters', len(text))

number of characters 1139


## [total points: 4] Exercise 1: NLTK
In this exercise, we use NLTK to apply **Part-of-speech (POS) tagging**, **Named Entity Recognition (NER)**, and **Constituency parsing**. The following code snippet already performs sentence splitting and tokenization. 

In [63]:
import nltk
from nltk.tokenize import sent_tokenize
from nltk import word_tokenize

In [64]:
sentences_nltk = sent_tokenize(text)

In [65]:
tokens_per_sentence = []
for sentence_nltk in sentences_nltk:
    sent_tokens = word_tokenize(sentence_nltk)
    tokens_per_sentence.append(sent_tokens)

We will use lists to keep track of the output of the NLP tasks. We can hence inspect the output for each task using the index of the sentence.

In [66]:
sent_id = 1
print('SENTENCE', sentences_nltk[sent_id])
print('TOKENS', tokens_per_sentence[sent_id])

SENTENCE The six phones and tablets affected are the Galaxy S III, running the new Jelly Bean system, the Galaxy Tab 8.9 Wifi tablet, the Galaxy Tab 2 10.1, Galaxy Rugby Pro and Galaxy S III mini.
TOKENS ['The', 'six', 'phones', 'and', 'tablets', 'affected', 'are', 'the', 'Galaxy', 'S', 'III', ',', 'running', 'the', 'new', 'Jelly', 'Bean', 'system', ',', 'the', 'Galaxy', 'Tab', '8.9', 'Wifi', 'tablet', ',', 'the', 'Galaxy', 'Tab', '2', '10.1', ',', 'Galaxy', 'Rugby', 'Pro', 'and', 'Galaxy', 'S', 'III', 'mini', '.']


### [point: 1] Exercise 1a: Part-of-speech (POS) tagging
Use `nltk.pos_tag` to perform part-of-speech tagging on each sentence.

Use `print` to **show** the output in the notebook (and hence also in the exported PDF!).

In [67]:
pos_tags_per_sentence = []

for tokens in tokens_per_sentence:
    pos_tags = nltk.pos_tag(tokens)
    pos_tags_per_sentence.append(pos_tags)
    print(pos_tags)
    print()

#loop over each stenence token and run postag

[('https', 'NN'), (':', ':'), ('//www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html', 'JJ'), ('Documents', 'NNS'), ('filed', 'VBN'), ('to', 'TO'), ('the', 'DT'), ('San', 'NNP'), ('Jose', 'NNP'), ('federal', 'JJ'), ('court', 'NN'), ('in', 'IN'), ('California', 'NNP'), ('on', 'IN'), ('November', 'NNP'), ('23', 'CD'), ('list', 'NN'), ('six', 'CD'), ('Samsung', 'NNP'), ('products', 'NNS'), ('running', 'VBG'), ('the', 'DT'), ('``', '``'), ('Jelly', 'RB'), ('Bean', 'NNP'), ("''", "''"), ('and', 'CC'), ('``', '``'), ('Ice', 'NNP'), ('Cream', 'NNP'), ('Sandwich', 'NNP'), ("''", "''"), ('operating', 'VBG'), ('systems', 'NNS'), (',', ','), ('which', 'WDT'), ('Apple', 'NNP'), ('claims', 'VBZ'), ('infringe', 'VB'), ('its', 'PRP$'), ('patents', 'NNS'), ('.', '.')]

[('The', 'DT'), ('six', 'CD'), ('phones', 'NNS'), ('and', 'CC'), ('tablets', 'NNS'), ('affected', 'VBN'), ('are', 'VBP'), ('the', 'DT'), ('Galaxy', 'NNP'), ('S', 'NNP'), ('III', 'NNP

In [68]:
print(pos_tags_per_sentence)

[[('https', 'NN'), (':', ':'), ('//www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html', 'JJ'), ('Documents', 'NNS'), ('filed', 'VBN'), ('to', 'TO'), ('the', 'DT'), ('San', 'NNP'), ('Jose', 'NNP'), ('federal', 'JJ'), ('court', 'NN'), ('in', 'IN'), ('California', 'NNP'), ('on', 'IN'), ('November', 'NNP'), ('23', 'CD'), ('list', 'NN'), ('six', 'CD'), ('Samsung', 'NNP'), ('products', 'NNS'), ('running', 'VBG'), ('the', 'DT'), ('``', '``'), ('Jelly', 'RB'), ('Bean', 'NNP'), ("''", "''"), ('and', 'CC'), ('``', '``'), ('Ice', 'NNP'), ('Cream', 'NNP'), ('Sandwich', 'NNP'), ("''", "''"), ('operating', 'VBG'), ('systems', 'NNS'), (',', ','), ('which', 'WDT'), ('Apple', 'NNP'), ('claims', 'VBZ'), ('infringe', 'VB'), ('its', 'PRP$'), ('patents', 'NNS'), ('.', '.')], [('The', 'DT'), ('six', 'CD'), ('phones', 'NNS'), ('and', 'CC'), ('tablets', 'NNS'), ('affected', 'VBN'), ('are', 'VBP'), ('the', 'DT'), ('Galaxy', 'NNP'), ('S', 'NNP'), ('III', 'NN

### [point: 1] Exercise 1b: Named Entity Recognition (NER)
Use `nltk.chunk.ne_chunk` to perform Named Entity Recognition (NER) on each sentence.

Use `print` to **show** the output in the notebook (and hence also in the exported PDF!).

In [69]:
from nltk.chunk import ne_chunk

ner_tags_per_sentence = []

#feed postagged output into ne_chunk() 
for pos_tags in pos_tags_per_sentence:
    ner_tree = ne_chunk(pos_tags)
    ner_tags_per_sentence.append(ner_tree)
    print(ner_tree)
    print()


(S
  https/NN
  :/:
  //www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html/JJ
  Documents/NNS
  filed/VBN
  to/TO
  the/DT
  (ORGANIZATION San/NNP Jose/NNP)
  federal/JJ
  court/NN
  in/IN
  (GPE California/NNP)
  on/IN
  November/NNP
  23/CD
  list/NN
  six/CD
  (ORGANIZATION Samsung/NNP)
  products/NNS
  running/VBG
  the/DT
  ``/``
  Jelly/RB
  (GPE Bean/NNP)
  ''/''
  and/CC
  ``/``
  Ice/NNP
  Cream/NNP
  Sandwich/NNP
  ''/''
  operating/VBG
  systems/NNS
  ,/,
  which/WDT
  (PERSON Apple/NNP)
  claims/VBZ
  infringe/VB
  its/PRP$
  patents/NNS
  ./.)

(S
  The/DT
  six/CD
  phones/NNS
  and/CC
  tablets/NNS
  affected/VBN
  are/VBP
  the/DT
  (ORGANIZATION Galaxy/NNP)
  S/NNP
  III/NNP
  ,/,
  running/VBG
  the/DT
  new/JJ
  (PERSON Jelly/NNP Bean/NNP)
  system/NN
  ,/,
  the/DT
  (ORGANIZATION Galaxy/NNP)
  Tab/NNP
  8.9/CD
  Wifi/NNP
  tablet/NN
  ,/,
  the/DT
  (ORGANIZATION Galaxy/NNP)
  Tab/NNP
  2/CD
  10.1/CD
  ,/,
  (P

In [70]:
print(ner_tags_per_sentence)

[Tree('S', [('https', 'NN'), (':', ':'), ('//www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html', 'JJ'), ('Documents', 'NNS'), ('filed', 'VBN'), ('to', 'TO'), ('the', 'DT'), Tree('ORGANIZATION', [('San', 'NNP'), ('Jose', 'NNP')]), ('federal', 'JJ'), ('court', 'NN'), ('in', 'IN'), Tree('GPE', [('California', 'NNP')]), ('on', 'IN'), ('November', 'NNP'), ('23', 'CD'), ('list', 'NN'), ('six', 'CD'), Tree('ORGANIZATION', [('Samsung', 'NNP')]), ('products', 'NNS'), ('running', 'VBG'), ('the', 'DT'), ('``', '``'), ('Jelly', 'RB'), Tree('GPE', [('Bean', 'NNP')]), ("''", "''"), ('and', 'CC'), ('``', '``'), ('Ice', 'NNP'), ('Cream', 'NNP'), ('Sandwich', 'NNP'), ("''", "''"), ('operating', 'VBG'), ('systems', 'NNS'), (',', ','), ('which', 'WDT'), Tree('PERSON', [('Apple', 'NNP')]), ('claims', 'VBZ'), ('infringe', 'VB'), ('its', 'PRP$'), ('patents', 'NNS'), ('.', '.')]), Tree('S', [('The', 'DT'), ('six', 'CD'), ('phones', 'NNS'), ('and', 'CC'),

### [points: 2] Exercise 1c: Constituency parsing
Use the `nltk.RegexpParser` to perform constituency parsing on each sentence.

Use `print` to **show** the output in the notebook (and hence also in the exported PDF!).

In [71]:
constituent_parser = nltk.RegexpParser('''
NP: {<DT>? <JJ>* <NN>*} # NP
P: {<IN>}           # Preposition
V: {<V.*>}          # Verb
PP: {<P> <NP>}      # PP -> P NP
VP: {<V> <NP|PP>*}  # VP -> V (NP|PP)*''')

In [72]:
constituency_output_per_sentence = []

for pos_tags in pos_tags_per_sentence:
    parse_tree = constituent_parser.parse(pos_tags)
    constituency_output_per_sentence.append(parse_tree)
    print(parse_tree)
    print()


(S
  (NP https/NN)
  :/:
  (NP
    //www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html/JJ)
  Documents/NNS
  (VP (V filed/VBN))
  to/TO
  (NP the/DT)
  San/NNP
  Jose/NNP
  (NP federal/JJ court/NN)
  (P in/IN)
  California/NNP
  (P on/IN)
  November/NNP
  23/CD
  (NP list/NN)
  six/CD
  Samsung/NNP
  products/NNS
  (VP (V running/VBG) (NP the/DT))
  ``/``
  Jelly/RB
  Bean/NNP
  ''/''
  and/CC
  ``/``
  Ice/NNP
  Cream/NNP
  Sandwich/NNP
  ''/''
  (VP (V operating/VBG))
  systems/NNS
  ,/,
  which/WDT
  Apple/NNP
  (VP (V claims/VBZ))
  (VP (V infringe/VB))
  its/PRP$
  patents/NNS
  ./.)

(S
  (NP The/DT)
  six/CD
  phones/NNS
  and/CC
  tablets/NNS
  (VP (V affected/VBN))
  (VP (V are/VBP) (NP the/DT))
  Galaxy/NNP
  S/NNP
  III/NNP
  ,/,
  (VP (V running/VBG) (NP the/DT new/JJ))
  Jelly/NNP
  Bean/NNP
  (NP system/NN)
  ,/,
  (NP the/DT)
  Galaxy/NNP
  Tab/NNP
  8.9/CD
  Wifi/NNP
  (NP tablet/NN)
  ,/,
  (NP the/DT)
  Galaxy/NNP

In [73]:
print(constituency_output_per_sentence)

[Tree('S', [Tree('NP', [('https', 'NN')]), (':', ':'), Tree('NP', [('//www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html', 'JJ')]), ('Documents', 'NNS'), Tree('VP', [Tree('V', [('filed', 'VBN')])]), ('to', 'TO'), Tree('NP', [('the', 'DT')]), ('San', 'NNP'), ('Jose', 'NNP'), Tree('NP', [('federal', 'JJ'), ('court', 'NN')]), Tree('P', [('in', 'IN')]), ('California', 'NNP'), Tree('P', [('on', 'IN')]), ('November', 'NNP'), ('23', 'CD'), Tree('NP', [('list', 'NN')]), ('six', 'CD'), ('Samsung', 'NNP'), ('products', 'NNS'), Tree('VP', [Tree('V', [('running', 'VBG')]), Tree('NP', [('the', 'DT')])]), ('``', '``'), ('Jelly', 'RB'), ('Bean', 'NNP'), ("''", "''"), ('and', 'CC'), ('``', '``'), ('Ice', 'NNP'), ('Cream', 'NNP'), ('Sandwich', 'NNP'), ("''", "''"), Tree('VP', [Tree('V', [('operating', 'VBG')])]), ('systems', 'NNS'), (',', ','), ('which', 'WDT'), ('Apple', 'NNP'), Tree('VP', [Tree('V', [('claims', 'VBZ')])]), Tree('VP', [Tree('V', [

Augment the RegexpParser so that it also detects Named Entity Phrases (NEP), e.g., that it detects *Galaxy S III* and *Ice Cream Sandwich*

In [74]:
constituent_parser_v2 = nltk.RegexpParser('''
NEP: {<NNP>+(<CD>|<NNP>|<NN>)*}   # NEP: proper nouns + optional nums/nouns
NP: {<DT>? <JJ>* <NN>*}            # NP
P: {<IN>}                           # Preposition
V: {<V.*>}                          # Verb
PP: {<P> <NP>}                      # PP -> P NP
VP: {<V> <NP|PP>*}                  # VP -> V (NP|PP)*''')


In [75]:
constituency_v2_output_per_sentence = []

for pos_tags in pos_tags_per_sentence:
    parse_tree_v2 = constituent_parser_v2.parse(pos_tags)
    constituency_v2_output_per_sentence.append(parse_tree_v2)
    print(parse_tree_v2)
    print()


(S
  (NP https/NN)
  :/:
  (NP
    //www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html/JJ)
  Documents/NNS
  (VP (V filed/VBN))
  to/TO
  (NP the/DT)
  (NEP San/NNP Jose/NNP)
  (NP federal/JJ court/NN)
  (P in/IN)
  (NEP California/NNP)
  (P on/IN)
  (NEP November/NNP 23/CD list/NN six/CD Samsung/NNP)
  products/NNS
  (VP (V running/VBG) (NP the/DT))
  ``/``
  Jelly/RB
  (NEP Bean/NNP)
  ''/''
  and/CC
  ``/``
  (NEP Ice/NNP Cream/NNP Sandwich/NNP)
  ''/''
  (VP (V operating/VBG))
  systems/NNS
  ,/,
  which/WDT
  (NEP Apple/NNP)
  (VP (V claims/VBZ))
  (VP (V infringe/VB))
  its/PRP$
  patents/NNS
  ./.)

(S
  (NP The/DT)
  six/CD
  phones/NNS
  and/CC
  tablets/NNS
  (VP (V affected/VBN))
  (VP (V are/VBP) (NP the/DT))
  (NEP Galaxy/NNP S/NNP III/NNP)
  ,/,
  (VP (V running/VBG) (NP the/DT new/JJ))
  (NEP Jelly/NNP Bean/NNP system/NN)
  ,/,
  (NP the/DT)
  (NEP Galaxy/NNP Tab/NNP 8.9/CD Wifi/NNP tablet/NN)
  ,/,
  (NP the/DT)
  (

In [76]:
print(constituency_v2_output_per_sentence)

[Tree('S', [Tree('NP', [('https', 'NN')]), (':', ':'), Tree('NP', [('//www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html', 'JJ')]), ('Documents', 'NNS'), Tree('VP', [Tree('V', [('filed', 'VBN')])]), ('to', 'TO'), Tree('NP', [('the', 'DT')]), Tree('NEP', [('San', 'NNP'), ('Jose', 'NNP')]), Tree('NP', [('federal', 'JJ'), ('court', 'NN')]), Tree('P', [('in', 'IN')]), Tree('NEP', [('California', 'NNP')]), Tree('P', [('on', 'IN')]), Tree('NEP', [('November', 'NNP'), ('23', 'CD'), ('list', 'NN'), ('six', 'CD'), ('Samsung', 'NNP')]), ('products', 'NNS'), Tree('VP', [Tree('V', [('running', 'VBG')]), Tree('NP', [('the', 'DT')])]), ('``', '``'), ('Jelly', 'RB'), Tree('NEP', [('Bean', 'NNP')]), ("''", "''"), ('and', 'CC'), ('``', '``'), Tree('NEP', [('Ice', 'NNP'), ('Cream', 'NNP'), ('Sandwich', 'NNP')]), ("''", "''"), Tree('VP', [Tree('V', [('operating', 'VBG')])]), ('systems', 'NNS'), (',', ','), ('which', 'WDT'), Tree('NEP', [('Apple', 'NN

## [total points: 1] Exercise 2: spaCy
Use Spacy to process the same text as you analyzed with NLTK.

In [77]:
import spacy
nlp = spacy.load('en_core_web_sm')

In [78]:
doc = nlp(text)
sents = list(doc.sents)


for i, sent in enumerate(sents):
    print("-"*100)
    print(f"\nSentence {i}:")
    print(sent.text)
    print()

    #pritn each token text, coarse POS, fg tag and dependency relation
    for token in sent:
        print(f"  {token.text:<20} pos: {token.pos_:<8} tag: {token.tag_:<6} dep: {token.dep_}")

    #filter doc level entitites dow nto only those in stenence token span
    ents_in_sent = [e for e in doc.ents if e.start >= sent.start and e.end <= sent.end]
    if ents_in_sent:
        print("\tNamed Entities:")
        for ent in ents_in_sent:
            print(f"\t\t{ent.text:<30} {ent.label_}")

    print()


----------------------------------------------------------------------------------------------------

Sentence 0:
https://www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html

Documents filed to the San Jose federal court in California on November 23 list six Samsung products running the "Jelly Bean" and "Ice Cream Sandwich" operating systems, which Apple claims infringe its patents.


  https://www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-products-under-scrutiny.html pos: PROPN    tag: NNP    dep: compound
  

                   pos: SPACE    tag: _SP    dep: dep
  Documents            pos: PROPN    tag: NNPS   dep: nsubj
  filed                pos: VERB     tag: VBD    dep: ROOT
  to                   pos: ADP      tag: IN     dep: prep
  the                  pos: DET      tag: DT     dep: det
  San                  pos: PROPN    tag: NNP    dep: compound
  Jose                 pos: PROPN    tag: NNP  

small tip: You can use **sents = list(doc.sents)** to be able to use the index to access a sentence like **sents[2]** for the third sentence.


In [84]:
#sentence splitting comparison

print(f"NLTK identified {len(sentences_nltk)} sentences")
print(f"spaCy identified {len(sents)} sentences")
print()

#zip compare if same same no same 
for i, (ns, ss) in enumerate(zip(sentences_nltk, sents)):
    match = "==" if ns.strip() == str(ss).strip() else "!="
    print(f"Sentence {i} {match}")
    #shorten to 90 chars bc no need to read allat 
    print(f"\tNLTK : {ns[:90]}")
    print(f"\tspaCy: {str(ss)[:90]}")
    print()


NLTK identified 6 sentences
spaCy identified 6 sentences

Sentence 0 ==
	NLTK : https://www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-produc
	spaCy: https://www.telegraph.co.uk/technology/apple/9702716/Apple-Samsung-lawsuit-six-more-produc

Sentence 1 ==
	NLTK : The six phones and tablets affected are the Galaxy S III, running the new Jelly Bean syste
	spaCy: The six phones and tablets affected are the Galaxy S III, running the new Jelly Bean syste

Sentence 2 !=
	NLTK : Apple stated it had “acted quickly and diligently" in order to "determine that these newly
	spaCy: Apple stated it had “acted quickly and diligently" in order to "determine that these newly

Sentence 3 !=
	NLTK : In August, Samsung lost a US patent case to Apple and was ordered to pay its rival $1.05bn
	spaCy: "
In August, Samsung lost a US patent case to Apple and was ordered to pay its rival $1.05

Sentence 4 ==
	NLTK : Samsung, which is the world's top mobile phone maker, is appealing t

In [ ]:
#NLTK keeps ." tgth as end of  stenece 2 while spacy treats " as the start of the next 
for i, (ns, ss) in enumerate(zip(sentences_nltk, sents)):
    if ns.strip() != str(ss).strip():
        print(f"Sentence {i} !=")
        print(f"\tNLTK : {ns}")
        print(f"\tspaCy: {str(ss)}")
        print()

Sentence 2 !=
	NLTK : Apple stated it had “acted quickly and diligently" in order to "determine that these newly released products do infringe many of the same claims already asserted by Apple."
	spaCy: Apple stated it had “acted quickly and diligently" in order to "determine that these newly released products do infringe many of the same claims already asserted by Apple.

Sentence 3 !=
	NLTK : In August, Samsung lost a US patent case to Apple and was ordered to pay its rival $1.05bn (£0.66bn) in damages for copying features of the iPad and iPhone in its Galaxy range of devices.
	spaCy: "
In August, Samsung lost a US patent case to Apple and was ordered to pay its rival $1.05bn (£0.66bn) in damages for copying features of the iPad and iPhone in its Galaxy range of devices.



In [87]:
#POS tag comparison for sentence 1 (NLTK vs spaCy)
nltk_tags  = pos_tags_per_sentence[1]
spacy_tags = [(token.text, token.tag_) for token in sents[1]]

print(f"{'Token':<20} {'NLTK':<10} {'spaCy':<10} {'Match?'}")
print("-" * 100)
for j in range(max(len(nltk_tags), len(spacy_tags))):
    n_word, n_tag = nltk_tags[j]  if j < len(nltk_tags)  else ("-", "-")
    s_word, s_tag = spacy_tags[j] if j < len(spacy_tags) else ("-", "-")
    #some positive vibes
    flag = "👍😁" if n_tag == s_tag else "!="
    print(f"{s_word:<20} {n_tag:<10} {s_tag:<10} {flag}")


Token                NLTK       spaCy      Match?
----------------------------------------------------------------------------------------------------
The                  DT         DT         👍😁
six                  CD         CD         👍😁
phones               NNS        NNS        👍😁
and                  CC         CC         👍😁
tablets              NNS        NNS        👍😁
affected             VBN        VBN        👍😁
are                  VBP        VBP        👍😁
the                  DT         DT         👍😁
Galaxy               NNP        NNP        👍😁
S                    NNP        NNP        👍😁
III                  NNP        NNP        👍😁
,                    ,          ,          👍😁
running              VBG        VBG        👍😁
the                  DT         DT         👍😁
new                  JJ         JJ         👍😁
Jelly                NNP        NNP        👍😁
Bean                 NNP        NNP        👍😁
system               NN         NN         👍😁
,                    

In [88]:
#NER comparison for sentence 3
sent_id = 3
print("NLTK NER (sentence 3)")
print(ner_tags_per_sentence[sent_id])
print()
print("spaCy NER (sentence 3)")
for ent in sents[sent_id].ents:
    print(f"  {ent.text:<30} {ent.label_}")


NLTK NER (sentence 3)
(S
  In/IN
  (GPE August/NNP)
  ,/,
  (PERSON Samsung/NNP)
  lost/VBD
  a/DT
  (GSP US/NNP)
  patent/NN
  case/NN
  to/TO
  (GPE Apple/NNP)
  and/CC
  was/VBD
  ordered/VBN
  to/TO
  pay/VB
  its/PRP$
  rival/JJ
  $/$
  1.05bn/CD
  (/(
  £0.66bn/NN
  )/)
  in/IN
  damages/NNS
  for/IN
  copying/VBG
  features/NNS
  of/IN
  the/DT
  (ORGANIZATION iPad/NN)
  and/CC
  (ORGANIZATION iPhone/NN)
  in/IN
  its/PRP$
  (GPE Galaxy/NNP)
  range/NN
  of/IN
  devices/NNS
  ./.)

spaCy NER (sentence 3)
  August                         DATE
  Samsung                        ORG
  US                             GPE
  Apple                          ORG
  1.05bn                         MONEY
  0.66bn                         MONEY
  iPad                           ORG
  iPhone                         ORG


In [89]:
#constituency  NLTK vs dep spaCy for sentence 3
sent_id = 3

print("NLTK Constituency Parse")
print(constituency_output_per_sentence[sent_id])
print()

print("spaCy Dependency Parse")
for token in sents[sent_id]:
    print(f"  {token.text:<15} --[{token.dep_:<10}]--> {token.head.text}")


NLTK Constituency Parse
(S
  (P In/IN)
  August/NNP
  ,/,
  Samsung/NNP
  (VP (V lost/VBD) (NP a/DT))
  US/NNP
  (NP patent/NN case/NN)
  to/TO
  Apple/NNP
  and/CC
  (VP (V was/VBD))
  (VP (V ordered/VBN))
  to/TO
  (VP (V pay/VB))
  its/PRP$
  (NP rival/JJ)
  $/$
  1.05bn/CD
  (/(
  (NP £0.66bn/NN)
  )/)
  (P in/IN)
  damages/NNS
  (P for/IN)
  (VP (V copying/VBG))
  features/NNS
  (PP (P of/IN) (NP the/DT iPad/NN))
  and/CC
  (NP iPhone/NN)
  (P in/IN)
  its/PRP$
  Galaxy/NNP
  (NP range/NN)
  (P of/IN)
  devices/NNS
  ./.)

spaCy Dependency Parse
  "               --[punct     ]--> lost
  
               --[dep       ]--> "
  In              --[prep      ]--> lost
  August          --[pobj      ]--> In
  ,               --[punct     ]--> lost
  Samsung         --[nsubj     ]--> lost
  lost            --[ROOT      ]--> lost
  a               --[det       ]--> case
  US              --[compound  ]--> case
  patent          --[compound  ]--> case
  case            --[dobj      ]--> lo

## [total points: 7] Exercise 3: Comparison NLTK and spaCy
We will now compare the output of NLTK and spaCy, i.e., in what do they differ?

### [points: 3] Exercise 3a: Part of speech tagging
Compare the output from NLTK and spaCy regarding part of speech tagging.

* To compare, you probably would like to compare sentence per sentence. Describe if the sentence splitting is different for NLTK than for spaCy. If not, where do they differ?
* After checking the sentence splitting, select a sentence for which you expect interesting results and perhaps differences. Motivate your choice.
* Compare the output in `token.tag` from spaCy to the part of speech tagging from NLTK for each token in your selected sentence. Are there any differences? This is not a trick question; it is possible that there are no differences.

### Answer

**Sentence splitting:**
Both NLTK and spaCy identified 6 sentences but disagree on the boundary between sentences 2 and 3.
NLTK keeps the closing quotation mark (`"`) attached to the end of sentence 2 ("...Apple."").
spaCy splits that closing quote off as a separate token and attaches it to the start of sentence 3.
All other sentence boundaries are identical.

**Selected sentence (index 1):**
*"The six phones and tablets affected are the Galaxy S III, running the new Jelly Bean
system, the Galaxy Tab 8.9 Wifi tablet, the Galaxy Tab 2 10.1, Galaxy Rugby Pro and
Galaxy S III mini."*

This sentence is a good test case because it contains product names with mixed
proper-noun/number tokens (Galaxy S III, Tab 8.9, Tab 2 10.1), which are tricky to tag.

**POS tag comparison:**
Both NLTK and spaCy use the Penn Treebank tagset, so tags are directly comparable.
Running the comparison on sentence 1, every single token receives the same tag from both tools.
The only difference is that spaCy produces an extra `_SP` (whitespace) token at the end of the sentence
that NLTK does not generate — NLTK simply ignores trailing whitespace.
This confirms that for straightforward product-name-heavy sentences, both taggers behave identically.


### [points: 2] Exercise 3b: Named Entity Recognition (NER)
* Describe differences between the output from NLTK and spaCy for Named Entity Recognition. Which one do you think performs better?

### Answer

**Differences between NLTK and spaCy NER sentence 3 as example:**

**NLTK output:**
NLTK misclassifies several entities: *August* is tagged as GPE (geopolitical entity????) *Samsung* is tagged as PERSON, *Apple* is tagged as GPE, and *Galaxy* is tagged as GPE all wrong. Pretty horrible overall.*iPad* and *iPhone* are tagged as ORGANIZATION which is debatable since they are products. On second thought its not really debatable at all its just wrong. *US* is tagged as GSP (geo spatial entity a subtype of GPE) which is alright for a country. Both monetary amounts (.05bn, £0.66bn) are missed entirely.

**spaCy output:**
spaCy correctly identifies: *Samsung* (ORG), *Apple* (ORG), *US* (GPE), *August* (DATE), *1.05bn* (MONEY), *0.66bn* (MONEY), *iPad* (ORG), *iPhone* (ORG). The only questionable label is ORG for iPad/iPhone they are products rather than orgs but NLTK did the same thing. 

**Verdict:** spaCy performs significantly better. It captures monetary values and dates that NLTK misses entirely and its entity type assignments are far more accurate. NLTK's rule-based chunker is limited by its POS input and coarse grammar rules.


### [points: 2] Exercise 3c: Constituency/dependency parsing
Choose one sentence from the text and run constituency parsing using NLTK and dependency parsing using spaCy.
* describe briefly the difference between constituency parsing and dependency parsing
* describe differences between the output from NLTK and spaCy.

### Answer

**Constituency vs Dependency parsing:**

*Constituency parsing* groups words into nested phrases (NP, VP, PP, S) forming a hierarchical tree. 
**what are the phrase-level building blocks of the sentence?**

*Dependency parsing* draws a directed arc from each word to its syntactic head labelled with a grammatical relation (nsubj, dobj, prep, conj, etc.).
**which word governs which other word, and how?**



#### Chosen sentence index 3:

**NLTK constituency output:**
The RegexpParser produces a shallow partial tree. Simple chunks are identified: (VP (V lost/VBD) (NP a/DT)), (NP patent/NN case/NN), (PP (P of/IN) (NP the/DT iPad/NN)). However the coordination "lost ... and was ordered" is not captured the grammarhas no rule for coordinated VPs so each verb appears in its own isolated VP chunk. Many tokens (August, Samsung, Apple, Galaxy) remain outside any phrase entirely.

**spaCy dependency output:**
Every token is linked to exactly one head with a labelled relation.
*lost* is the ROOT. 
*Samsung* is its nsubj, 
*case* is its dobj,
*In* is a prep modifier.
*ordered* is joined to *lost* via conj, capturing the coordination.
*pay* is the xcomp of *ordered*, and *rival* is the dobj of *pay*.
Nested PPs ("in damages", "for copying", "of features", "of range") are all linked correctly.

**TLDR:** NLTK produces a partial flat chunking limited by the hand-written rules spaCy produces a complete dependency graph covering every token in the sentence.


# End of this notebook